test/e2e/subroutes_test.go

```diff
@@ -109,66 +232,155 @@ func TestSubrouteVisibilityChange(t *testing.T) {
 	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
 	defer test.TearDown(clients, names)
 
-	tag := "my-tag"
+	subrouteTag1 := "my-tag"
+	subrouteTag2 := "my-tag2"
 
-	withInternalVisibility := WithServiceLabel(
-		routeconfig.VisibilityLabelKey, routeconfig.VisibilityClusterLocal)
+	withInternalVisibility := WithServiceLabel(routeconfig.VisibilityLabelKey, routeconfig.VisibilityClusterLocal)
 	withTrafficSpec := WithInlineRouteSpec(v1alpha1.RouteSpec{
 		Traffic: []v1alpha1.TrafficTarget{
 			{
 				TrafficTarget: v1.TrafficTarget{
-					Tag:     tag,
-					Percent: ptr.Int64(100),
+					Tag:     subrouteTag1,
+					Percent: ptr.Int64(50),
+				},
+			},
+			{
+				TrafficTarget: v1.TrafficTarget{
+					Tag:     subrouteTag2,
+					Percent: ptr.Int64(50),
 				},
 			},
 		},
 	})
-	resources, err := v1a1test.CreateRunLatestServiceReady(t, clients, &names, withInternalVisibility, withTrafficSpec)
+	resources, err := v1a1test.CreateRunLatestServiceReady(t, clients, &names, withTrafficSpec, withInternalVisibility)
 	if err != nil {
 		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
 	}
 
-	// Ensure that it only has cluster local addresses
-	for _, traffic := range resources.Route.Status.Traffic {
-		if !strings.HasSuffix(traffic.TrafficTarget.URL.Host, network.GetClusterDomainName()) {
-			t.Fatalf("Expected all subroutes to be cluster local")
+	for _, tag := range []string{subrouteTag1, subrouteTag2} {
+		if isClusterLocal, err := trafficWithTagIsClusterLocal(resources.Route.Status.Traffic, tag); err != nil {
+			t.Fatalf(err.Error())
+		} else if !isClusterLocal {
+			t.Fatalf("Expected subroute with tag %s to be cluster local", tag)
 		}
 	}
 
-	serviceName := fmt.Sprintf("%s-%s", tag, resources.Route.Name)
+	if !isRouteClusterLocal(resources.Route.Status) {
+		t.Fatalf("Expected route to be cluster local")
+	}
+
+	// change subroute 1 - private
+	serviceName := fmt.Sprintf("%s-%s", subrouteTag1, resources.Route.Name)
 	svc, err := clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Get(serviceName, metav1.GetOptions{})
 	if err != nil {
 		t.Fatalf("Failed to get k8s service to modify: %s", err.Error())
 	}
 
 	svcCopy := svc.DeepCopy()
-	labels.SetVisibility(&svcCopy.ObjectMeta, false)
+	labels.SetVisibility(&svcCopy.ObjectMeta, true)
+
+	svcpatchBytes, err := test.CreateBytePatch(svc, svcCopy)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	if _, err = clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Patch(serviceName, types.JSONPatchType, svcpatchBytes); err != nil {
+		t.Fatalf("Failed to patch service: %s", err.Error())
+	}
 
-	patch, err := duck.CreatePatch(svc, svcCopy)
+	// TODO: Add a better check here. ex. Wait 2s checking every 100ms to ensure it doesn't get the wrong visibility
+	time.Sleep(2 * time.Second) // TODO: Remove this after better check ^^
```

Comment:
Would it make sense to do what you are doing for `TestSubrouteVisibilityPublicToPrivate` to update the route spec to ensure that route reconciliation is completed before the upcoming checks?

Reply:
Test `TestSubrouteVisibilityPublicToPrivate ` checks that route has met certain condition in `Wait` function but in this test case we need to consistently check for sub-route to remain private. Thats why the `sleep` is added and then traffic is checked for cluster local. I have updated this logic, please review and let me know if this looks fine.